src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableList_1.cs

```diff
@@ -2205,19 +2205,22 @@ internal Node RemoveAll(Predicate<T> match)
                 Contract.Ensures(Contract.Result<Node>() != null);
 
                 var result = this;
-                int index = 0;
-                foreach (var item in this)
-                {
+                var index = 0;
+                var count = this.Count;
+                var removed = 0;
+                for(;;)
+                {
+                    if (index + removed >= count)
+                        break;
+                    var item = result[index];
```

Comment:
OK... it is certainly possible to do this without a n-log-n best case, and I appreciate that point of view.  The complexity of the code would be considerably worse, I think, due to the fact that you need a way to basically the same thing the iterator is doing (either using it or using separate code) and a way to sync that up with the new result when the nodes are mutated.  I can think about what that might look like.  Are you opposed to a public method on the Enumerator that would allow it to "adapt" to a removal of an item?